jinja_loader type bug

I have
```Python
from jinja2 import ChoiceLoader, FileSystemLoader


app.jinja_loader = ChoiceLoader([
    FileSystemLoader(project_dir / 'templates/'),
    app.jinja_loader,
])
```

There is a mypy error
```
error: Incompatible types in assignment (expression has type "ChoiceLoader", variable has type "FileSystemLoader | None")  [assignment]
error: List item 1 has incompatible type "FileSystemLoader | None"; expected "BaseLoader"  [list-item]
```

There shouldn't be an error.

Environment:

- Python version: 3.12.1
- Flask version: 3.0.1

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -3,6 +3,8 @@ Version 3.0.2
 
 Unreleased
 
+-   Correct type for ``jinja_loader`` property. :issue:`5388`
+
 
 Version 3.0.1
 -------------

--- a/src/flask/sansio/scaffold.py
+++ b/src/flask/sansio/scaffold.py
@@ -9,6 +9,7 @@ from collections import defaultdict
 from functools import update_wrapper
 
 import click
+from jinja2 import BaseLoader
 from jinja2 import FileSystemLoader
 from werkzeug.exceptions import default_exceptions
 from werkzeug.exceptions import HTTPException
@@ -272,7 +273,7 @@ class Scaffold:
         self._static_url_path = value
 
     @cached_property
-    def jinja_loader(self) -> FileSystemLoader | None:
+    def jinja_loader(self) -> BaseLoader | None:
         """The Jinja loader for this object's templates. By default this
         is a class :class:`jinja2.loaders.FileSystemLoader` to
         :attr:`template_folder` if it is set.

